### DISCUSSION TITLE
Stripe webhooks – `uninitialized constant FulfillCheckout` error on `Pay::Webhooks.delegator.subscribe`

### DISCUSSION THREAD

**Discussion by rameerez** [pay-rails/pay#944] [Q&A] [answered]
Hi everyone, I'm following the docs and trying to configure custom Stripe webhook listeners as described [here](https://github.com/pay-rails/pay/blob/main/docs/stripe/8_stripe_checkout.md)

I add the `Pay::Webhooks.delegator.subscribe` lines to `config/initializers/pay.rb`

And then I create a `app/webhooks/charge_succeeded.rb` file where I put the `class FulfillCheckout` definition.

However, on `rails s`, the app crashes with the following error:
```
Exiting
myapp/config/initializers/pay.rb:1:in `<main>': uninitialized constant FulfillCheckout (NameError)

Pay::Webhooks.delegator.subscribe "stripe.checkout.session.completed", FulfillCheckout.new
```

I tried adding `config.autoload_paths << Rails.root.join('app/webhooks')` to `config/application.rb` in case the custom `app/webhooks` folder wasn't getting auto-loaded but that doesn't work either – and in any case Zeitwerk should autoload everything under `app`, right?

As suggested in #486, I also tried this lambda approach:
```ruby
Pay::Webhooks.delegator.subscribe "stripe.charge.succeeded", ->(event) { FulfillCheckout.new.call(event) }
```

– but no luck either, same uninitialized constant error.

I've also seen people defining the `FulfillCheckout` lambda/class itself inside `config/initializers/pay.rb` as in #517, but that doesn't seem scalable as I start adding more custom webhooks.

Also tried matching the name of the class to the name of the file (`charge_succeeded.rb` -> `class ChargeSucceeded`) but I get the exact same error all the time.

It looks like Rails is just not loading my `app/webhooks` folder – any ideas?

Using ruby 3.2.2 on rails 7.1.2

**Comment by deanpcmad** [comment]
Your issue is likely due to the file name you've chosen as Zeitwerk, the autoloader that Rails uses, is expecting your `charge_succeeded.rb` file to have the class `ChargeSucceeded`

From the [Rails docs](https://guides.rubyonrails.org/v6.0.3/autoloading_and_reloading_constants.html#project-structure):

>  Rails configures Zeitwerk to inflect file names with String#camelize. For example, it expects that app/controllers/users_controller.rb defines the constant UsersController 

If you rename the `charge_succeeded.rb` file to `fulfill_checkout.rb`, then it will likely work as expected.

**Reply by rameerez** [reply]
Already tried, doesn't work either. Same error.

**Comment by rameerez** [comment]
Update: looks like it works if I manually require each file in `config/initializers/pay.rb`, as in:
```ruby
require Rails.root.join('app/webhooks/fulfill_checkout.rb')

Pay::Webhooks.delegator.subscribe "stripe.checkout.session.completed", FulfillCheckout.new
Pay::Webhooks.delegator.subscribe "stripe.checkout.session.async_payment_succeeded", FulfillCheckout.new
```

Not sure this is the best solution, though, since Zeitwerk should be doing this work.

**Reply by deanpcmad** [reply]
Yeah I was just writing out a comment where I found the same issue in my Pay test app. I did manage to get it to work by prepending `Application::` to the `FulfillCheckout` class but then that stopped working after a few tries 🤔 

If you comment it out, the `FulfillCheckout` class is available in the console, so Zeitwerk is loading it correctly, but it seems that it doesn't load it before the initializers? Very odd.

**Reply by deanpcmad** [reply]
Also, side note, the PR #943 might be of interest for you which Chris has been working on from discussion #938

**Comment by rameerez** [comment, marked as answer]
I found out what I was doing wrong: the `Pay::Webhooks.delegator.subscribe` lines need to go inside an `ActiveSupport.on_load(:pay)` block, like this:

```ruby
# config/initializers/pay.rb

ActiveSupport.on_load(:pay) do
  Pay::Webhooks.delegator.subscribe "stripe.checkout.session.completed", FulfillCheckout.new
end
```

Doing this, there's no need to manually require any files.

Turns out this was not mentioned in the `pay` docs for Stripe Checkout ([`docs/stripe/8_stripe_checkout.md`](https://github.com/pay-rails/pay/blob/main/docs/stripe/8_stripe_checkout.md)) – but it was mentioned in the regular webhooks docs ([`docs/7_webhooks.md`](https://github.com/pay-rails/pay/blob/main/docs/7_webhooks.md))

Looks like [this change was introduced in version `6.4.0`](https://github.com/pay-rails/pay/blob/main/CHANGELOG.md#640) and the Stripe docs were not updated. I'm submitting a PR to fix the Stripe Checkout docs.
